src/DataSourcePoller.php

```diff
@@ -0,0 +1,231 @@
+<?php
+
+namespace Automattic\WooCommerce\Admin;
+
+/**
+ * Specs data source poller class.
+ * This handles polling specs from JSON endpoints, and
+ * stores the specs in to the database as an option.
+ */
+class DataSourcePoller {
+
+	/**
+	 * Name of data sources filter.
+	 */
+	const FILTER_NAME = 'data_source_poller_data_sources';
+
+	/**
+	 * Default data sources array.
+	 *
+	 * @var array
+	 */
+	protected $data_sources = array();
+
+	/**
+	 * Key for the spec id.
+	 *
+	 * @var string
+	 */
+	protected $spec_key = 'id';
+
+	/**
+	 * Key for the spec id.
+	 *
+	 * @var string
+	 */
+	protected $transient_name = '';
+
+	/**
+	 * The logger instance.
+	 *
+	 * @var WC_Logger|null
+	 */
+	protected static $logger = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param array  $data_sources urls for data sources.
+	 * @param string $transient_name name of transient.
+	 * @param string $spec_key Optional key used as the spec identifier.
+	 */
+	public function __construct( $data_sources, $transient_name, $spec_key = 'id' ) {
+		$this->data_sources   = $data_sources;
+		$this->transient_name = $transient_name;
+		$this->spec_key       = $spec_key;
+	}
+
+	/**
+	 * Get the logger instance.
+	 *
+	 * @return WC_Logger
+	 */
+	protected static function get_logger() {
+		if ( is_null( self::$logger ) ) {
+			self::$logger = wc_get_logger();
+		}
+
+		return self::$logger;
+	}
+
+	/**
+	 * Returns the key identifier of spec, this can easily be overwritten. Defaults to id.
+	 *
+	 * @param mixed $spec a JSON parsed spec coming from the JSON feed.
+	 * @return string|boolean
+	 */
+	protected function get_spec_key( $spec ) {
+		$key = $this->spec_key;
+		if ( isset( $spec->$key ) ) {
+			return $spec->$key;
+		}
+		return false;
+	}
+
+	/**
+	 * Reads the data sources for specs and persists those specs.
+	 *
+	 * @return array list of specs.
+	 */
+	public function get_specs_from_data_sources() {
+		$specs = get_transient( $this->transient_name );
+
+		if ( false === $specs || ! is_array( $specs ) || 0 === count( $specs ) ) {
+			$this->read_specs_from_data_sources();
+			$specs = get_transient( $this->transient_name );
+		}
+		return false !== $specs ? $specs : array();
+	}
+
+	/**
+	 * Reads the data sources for specs and persists those specs.
+	 *
+	 * @return bool Whether any specs were read.
+	 */
+	public function read_specs_from_data_sources() {
+		$specs        = array();
+		$data_sources = apply_filters( self::FILTER_NAME, $this->data_sources );
+
+		// Note that this merges the specs from the data sources based on the
+		// id - last one wins.
+		foreach ( $data_sources as $url ) {
+			$specs_from_data_source = self::read_data_source( $url );
+			$this->merge_specs( $specs_from_data_source, $specs, $url );
+		}
+
+		// Persist the specs as a transient.
+		set_transient(
+			$this->transient_name,
+			$specs,
+			7 * DAY_IN_SECONDS
```

Comment:
Optional, but might be nice to allow control of this from extending classes.

Reply:
Yeah good idea.